src/ZoneServer/World/Actors/Characters/Components/PersonalStorageComponent.cs

```diff
@@ -0,0 +1,512 @@
+					}
+				}
+			}
+
+			return StorageResult.InvalidOperation;
+		}
+
+		/// <summary>
+		/// Removes an item from storage.
+		/// Does not update client.
+		/// </summary>
+		/// <param name="item">Item to remove</param>
+		/// <param name="amount">Amount of the item</param>
+		/// <param name="removedPosition">Position where item was removed. Defaults to -1</param>
+		/// <param name="removedAmount">Amount that was removed. Defaults to 0</param>
+		/// <returns></returns>
+		public StorageResult Remove(Item item, int amount, out int removedPosition, out int removedAmount)
+		{
+			removedAmount = 0;
+			removedPosition = -1;
+
+			if (amount <= 0)
+				return StorageResult.InvalidOperation;
+
+			var result = this.TryGetItem(item.ObjectId, out item, out removedPosition);
+			if (result != StorageResult.Success)
+				return result;
+
+			// If the item is stackable, handle the reduction of the stack
+			if (item.IsStackable)
+			{
+				lock (_syncLock)
+					removedAmount = RemoveItemStack(item, amount, removedPosition);
+			}
+			else
+			{
+				lock (_syncLock)
+				{
+					_storageItems[removedPosition] = null;
+					_occupiedPositions -= 1;
+					removedAmount = 1;
+				}
+			}
+			return StorageResult.Success;
+		}
+
+		/// <summary>
+		/// Returns true if storage is full.
+		/// </summary>
+		/// <returns></returns>
+		public bool CheckStorageFull()
+		{
+			lock (_syncLock)
+				return _occupiedPositions == _maxStorage;
+		}
+
+		/// <summary>
+		/// Returns a dictionary of items in each position
+		/// of storage.
+		/// </summary>
+		/// <returns></returns>
+		public Dictionary<int, Item> GetStorage()
+		{
+			lock (_syncLock)
+			{
+				// Filter out null items and create a dictionary
+				var storageDictionary = _storageItems
+					.Select((item, index) => new { Item = item, Index = index })
+					.Where(itemData => itemData.Item != null)
+					.ToDictionary(itemData => itemData.Index, itemData => itemData.Item);
+
+				return storageDictionary;
+			}
+		}
+
+		/// <summary>
+		/// Returns specific item, or null if it doesn't exist.
+		/// </summary>
+		/// <param name="worldId">Item object id</param>
+		/// <param name="item">Item found</param>
+		/// <param name="position">Position of item found in storage</param>
+		/// <returns></returns>
+		public StorageResult TryGetItem(long worldId, out Item item, out int position)
+		{
+			item = null;
+			position = -1;
+
+			lock (_syncLock)
+			{
+				for (var i = 0; i < _maxStorage; i++)
+				{
+					if ((_storageItems[i] != null) && (_storageItems[i].ObjectId == worldId))
+					{
+						item = _storageItems[i];
+						position = i;
+						return StorageResult.Success;
+					}
+				}
+			}
+
+			return StorageResult.ItemNotFound;
+		}
+
+		/// <summary>
+		/// Returns specific item in position, or null
+		/// if it doesn't exist.
+		/// </summary>
+		/// <param name="position"></param>
+		/// <returns></returns>
+		public Item TryGetItemPosition(int position)
```

Comment:
Minor thing, but adding an "At" in between Item and Position would more accurately describe what this method does.